client/apis/volumesnapshot/v1/types.go

```diff
@@ -0,0 +1,429 @@
+type VolumeSnapshotSource struct {
+	// persistentVolumeClaimName specifies the name of the PersistentVolumeClaim
+	// object representing the volume from which a snapshot should be created.
+	// This PVC is assumed to be in the same namespace as the VolumeSnapshot
+	// object.
+	// This field should be set if the snapshot does not exists, and should be
+	// created.
+	// This field is immutable.
+	// +optional
+	PersistentVolumeClaimName *string `json:"persistentVolumeClaimName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+	// volumeSnapshotContentName specifies the name of a pre-existing VolumeSnapshotContent
+	// object representing an existing volume snapshot.
+	// This field should be set if the snapshot already exists.
+	// This field is immutable.
+	// +optional
+	VolumeSnapshotContentName *string `json:"volumeSnapshotContentName,omitempty" protobuf:"bytes,2,opt,name=volumeSnapshotContentName"`
+}
+
+// VolumeSnapshotStatus is the status of the VolumeSnapshot
+// Note that CreationTime, RestoreSize, ReadyToUse, and Error are in both
+// VolumeSnapshotStatus and VolumeSnapshotContentStatus. Fields in VolumeSnapshotStatus
+// are updated based on fields in VolumeSnapshotContentStatus. They are eventual
+// consistency. These fields are duplicate in both objects due to the following reasons:
+// - Fields in VolumeSnapshotContentStatus can be used for filtering when importing a
+//   volumesnapshot.
+// - VolumsnapshotStatus is used by end users because they cannot see VolumeSnapshotContent.
+// - CSI snapshotter sidecar is light weight as it only watches VolumeSnapshotContent
+//   object, not VolumeSnapshot object.
+type VolumeSnapshotStatus struct {
+	// boundVolumeSnapshotContentName is the name of the VolumeSnapshotContent
+	// object to which this VolumeSnapshot object intends to bind to.
+	// If not specified, it indicates that the VolumeSnapshot object has not been
+	// successfully bound to a VolumeSnapshotContent object yet.
+	// NOTE: To avoid possible security issues, consumers must verify binding between
+	// VolumeSnapshot and VolumeSnapshotContent objects is successful (by validating that
+	// both VolumeSnapshot and VolumeSnapshotContent point at each other) before using
+	// this object.
+	// +optional
+	BoundVolumeSnapshotContentName *string `json:"boundVolumeSnapshotContentName,omitempty" protobuf:"bytes,1,opt,name=boundVolumeSnapshotContentName"`
+
+	// creationTime is the timestamp when the point-in-time snapshot is taken
+	// by the underlying storage system.
+	// In dynamic snapshot creation case, this field will be filled in by the
+	// snapshot controller with the "creation_time" value returned from CSI
+	// "CreateSnapshot" gRPC call.
+	// For a pre-existing snapshot, this field will be filled with the "creation_time"
+	// value returned from the CSI "ListSnapshots" gRPC call if the driver supports it.
+	// If not specified, it indicates that the creation time of the snapshot is unknown.
```

Comment:
nit:
```suggestion
	// If not specified, it may indicate that the creation time of the snapshot is unknown.
```

Could also mean controller just hasn't gotten around to updating it yet.